Alasdair/multi asset

Ali-Hill opened the pull request:
Adding Multi Asset support to the ledger model.

WhatisRT reviewed (changes requested):
We don't have a style guide yet so for now I'm happy with whatever whitespace/formatting, except that I'd like changes to be intentional and I'd generally try to have certain things aligned and avoid long lines.

With that in mind, maybe go over the diff once and see which changes you'd like to keep and remove the rest? There are also some leftover temporary comments and files that should be removed.

I'll take a quick look and see if I can fix the issue `deriveComputational` has with the equation for `mint`.

WhatisRT commented:
The issue with the equation is that `tryConstrsWith'` in `reduceDec` runs out of fuel (it's currently hard-coded to 5). It needs one fuel per clause, and this is simply adding a sixth. You could just set it to 10 or so for now. I've opened an issue for this: #48

Ali-Hill commented on src/Ledger/TokenAlgebra.lagda:
```diff
@@ -0,0 +1,36 @@
+  field property : coin ∘ inject ≗ id
+
+
+  field relImpliesCoinEquality : ∀{v1 v2} → v1 ≈ v2 → coin v1 ≡ coin v2
```
Yeah it is

Ali-Hill commented on src/Ledger/Transaction.lagda:
```diff
@@ -90,22 +101,27 @@ This function must produce a unique id for each unique transaction body.
 
   record TxWitnesses : Set where
     field vkSigs   : VKey ↛ Sig
-          scripts  : ℙ Script
```
This wasn't changed by me. This is what was is already in the master branch

Ali-Hill commented on src/Ledger/Utxo/Properties.lagda:
```diff
@@ -44,18 +46,50 @@ private variable
 newTxid⇒disj id∉utxo = disjoint⇒disjoint' λ h h' → id∉utxo $ to ∈-map
   (-, (case from ∈-map h' of λ where (_ , refl , h'') → case from ∈-map h'' of λ where (_ , refl , _) → refl) , h)
+
+{-
```
There aren't any unused lemmas. I commented out lemmas that are needed when minting was enabled again

WhatisRT commented:
There are still a bunch of redundant changes and files in this PR, as well as changes that look unintentional to me. Can you please go over the diff and remove those?

Ali-Hill commented:
I have gone over the diff and removed any of the redundant changes I saw. The only thing I have added that I don't use is the TotalMap definitions. Should I remove this as well?

WhatisRT commented on src/Axiom/Set/TotalMap.agda:
```diff
@@ -0,0 +1,40 @@
+open import Interface.DecEq
+
+private variable A A' B : Type
+                 R R' : Rel A B
```
I think you don't need all these variables

WhatisRT commented on src/Axiom/Set/TotalMap.agda:
```diff
@@ -0,0 +1,40 @@
+open import Axiom.Set
+
+module Axiom.Set.TotalMap (th : Theory) where
+open Theory th
```
Adding an empty line after `module` makes it easier to see. Also, do you need all those imports? Grouping them and sorting them alphabetically would also be nice.

WhatisRT commented on src/Ledger/Foreign/HSLedger.agda:
```diff
@@ -88,6 +88,27 @@ HSScriptStructure : ScriptStructure ℕ ℕ ℕ
+
+coinTokenAlgebra : TokenAlgebra
+coinTokenAlgebra = record
+         { PolicyId = ℕ
```
Please align the `=` signs and change the indentation to 2 spaces

WhatisRT commented on src/Ledger/Foreign/HSLedger.agda:
```diff
@@ -88,6 +88,27 @@ HSScriptStructure : ScriptStructure ℕ ℕ ℕ
+         ; property = λ x → refl
+         ; coinIsMonoidMorphism = record { mn-homo = record
+                                                     { sm-homo = record { ⟦⟧-cong = λ z → z
+                                                       ; ∙-homo = λ x y → refl }
```
Please align the `;`'s with the `{` above and change the indentation to two spaces

WhatisRT commented on src/Ledger/Foreign/HSLedger.agda:
```diff
@@ -138,16 +162,18 @@ instance
-        ; txsize   = txsize
-        ; txid     = txid
-        }
+                    { txins = from ⦃ Convertible-FinSet ⦃ Coercible⇒Convertible ⦄ ⦄ txins
```
Please align the `=` signs and restore the old indentation

WhatisRT commented on src/Ledger/Foreign/HSLedger.agda:
```diff
@@ -188,22 +214,26 @@ instance
-        ; Emax          = Emax
-        ; pv            = coerce pv
-        }
+                   { a = a
```
Same as above, `=` signs and indentation

WhatisRT commented on src/Ledger/Transaction.lagda:
```diff
@@ -12,10 +12,15 @@ open import Ledger.Prelude
 import Data.Nat as N
+open import Data.Integer
+
+data Tag : Set where
```
This type should be unnecessary

WhatisRT commented on src/Ledger/Utxo.lagda:
```diff
@@ -21,19 +21,34 @@ open import MyDebugOptions
+utxoEntrySize : TxOut → MemoryEstimate
+utxoEntrySize (fst , v) = utxoEntrySizeWithoutVal + size v
+
+-- fix this
```
```suggestion
-- TODO: fix this
```

WhatisRT commented on src/Ledger/Utxo.lagda:
```diff
@@ -151,26 +182,35 @@ data _⊢_⇀⦇_,UTXO⦈_ where
           fees = UTxOState.fees s
       in
-    txins tx ≢ ∅
+     txins tx ≢ ∅
```
```suggestion
    txins tx ≢ ∅
```

WhatisRT commented on src/Ledger/Utxo.lagda:
```diff
@@ -151,26 +182,35 @@ data _⊢_⇀⦇_,UTXO⦈_ where
       ⟦ (utxo ∣ txins tx ᶜ) ∪ᵐˡ outs tx , fees + f ⟧ᵘ
 \end{code}
 \begin{code}[hide]
--- TODO: This can't be moved into Properties because it breaks. Move
```
Restore this TODO

WhatisRT commented:
Keeping the `TotalMap` is fine, since we need that to actually implement MA

WhatisRT commented:
I've rebased this on top of the branch that has CI

jmchapman commented:
is this blocked by PR #51 ?

jmchapman commented:
or are there more changes to be made?

williamdemeo commented:
I'm working on this branch now.

@WhatisRT made a ticket (#56) for the missing MA stuff and assigned it to me. 

Andre said he also rebased the MA branch, so the only thing in the diff that doesn't belong there is the CI PR (which he'll merge tomorrow when the action has finished, and then that stuff will also disappear).

Andre also said he has already reviewed this, but since I'll be building right on top of it, it's probably a good starting point.

williamdemeo commented:
For this branch, the `nix-build` commands,
```
nix-build -A ledger.executableSpec    
nix-build -A ledger.docs
nix-build -A midnight.executableSpec
nix-build -A midnight.docs
```

described in the README.md file work fine, but if I load `src/Everything.agda` into emacs and try to type-check, I get an 8700 line error, the start of which is

```
.../formal-ledger-specifications/src/Ledger/Utxo.lagda:221,1-94
reduceDec: Could not find an equation to rewrite with!
when checking that the expression
unquote
(by-derive⇔ (quote _⊢_⇀⦇_,UTXO⦈_)
 (_∷_ {_} {_}
  (Tactic.DeriveComp.STSConstr.constructor (quote UTXO-inductive)
   (takeWhile {zeroˡ}
    {_$_ {sucˡ zeroˡ} {sucˡ zeroˡ} {Set₀} {λ z → Set₀}
     (Agda.Builtin.Reflection.Abs {zeroˡ})
     (Agda.Builtin.Reflection.Arg {zeroˡ} Agda.Builtin.Reflection.Type)}
    {zeroˡ}
    {λ z →
```

WhatisRT commented:
That error is the tactic running out of fuel, which is fixed by rebuilding Agda with its dependencies.

From my POV this PR is now finished but I'll wait for you to review it. Since it also has a messy history, I'll squash merge it once you've reviewed it.

williamdemeo reviewed (approved):
This looks great.  Nice job Alasdair.  I can't say I fully understand everything yet, but I'm getting closer... and at least I can confirm that everything type-checks on both macos (darwin) and ubuntu (22.04). 👍🏼
